docs/root/intro/arch_overview/security/ssl.rst

```diff
@@ -114,21 +114,35 @@ Certificate selection
 
-* Only one certificate of a particular type (RSA or ECDSA) may be specified.
+Certificate config/loading rules:
+
+* DNS SANs or Subject Common Name is extracted as server name pattern to match SNI during handshake.
+* FQDN like "test.example.com" and wildcard like "\*.example.com" are valid at the same time, which will be loaded
+  as two different server name pattern.
+* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
 * Non-P-256 server ECDSA certificates are rejected.
-* If the client supports P-256 ECDSA, a P-256 ECDSA certificate will be selected if one is present in the
-  :ref:`DownstreamTlsContext <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`
-  and it is in compliance with the OCSP policy.
-* If the client only supports RSA certificates, a RSA certificate will be selected if present in the
-  :ref:`DownstreamTlsContext <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.
-* Otherwise, the first certificate listed is used. This will result in a failed handshake if the
-  client only supports RSA certificates and the server only has ECDSA certificates.
 * Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
   <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.
-* The selected certificate must adhere to the OCSP policy. If no
-  such certificate is found, the connection is refused.
+
+Certificate selection rules, including SNI matching, Public Key Type matching and OCSP:
+
+* If the client supports SNI, a certificate with proper DNS SANs or Subject Common Name should be selected
+* It matches on exact server name first, then matches on wildcard server name if an exact name match isn't found, e.g. if SNI is
+  "test.example.com", a certificate with "test.example.com" will be preferred over for "\*.example.com".
+* If no certificate is matched to SNI or the client does not support SNI, subsequent particular type (RSA or ECDSA)
+  matching will be executed with all certificates as candidates.
+* Otherwise, particular type (RSA or ECDSA) matching will be executed with SNI-matched certificates as candidates.
+* Key type matching is based on SNI matching results
+* If the client supports P-256 ECDSA, the first P-256 ECDSA certificate is selected if it is present and the OCSP check is passed.
+* If the client only supports RSA, the first RSA certificate is selected if it is present.
+* If no exact match, fallback to the first certificate in the candidates.
+* The certificate that it fallbacks to might result in a failed handshake. For instance, a client only supports
+  RSA certificates and the certificate only support ECDSA, or a client only supports ECDSA certificate and the
+  certificate only support RSA.
+* After key type matching, one certificate is selected, and the selected certificate must adhere to the OCSP policy.
```

Comment:
This is a change from the previous behavior. I think you need to retain previous behavior. The previous docs specify that a cert is only selected "if it is in compliance with the OCSP policy." 

To be consistent, we should probably check this while doing SNI matching, so that if test.example.com cert is out of OCSP policy but *.example.com is in policy, *.example.com is selected.

Reply:
we do not break previous docs, currently if ocsp fails, we return ssl_select_cert_error.

I think fallback to wildcard SNI matching due to OCSP fail brings some complexity, what if pkey matching fails? Do we need to look up wildcard SNI matching again. Which cert should we fallback to, the first exact SNI-matched certs, the first wildcard SNI-matched certs, or first cert without any match?

Reply:
I wrote alternative code to support wildcard SNI matching if ocsp/pkey matching fails with exact SNI, and we will always fallback to the first cert in the tls context list.

```cpp
  auto selected = [&](const TlsContext& ctx) -> bool {
    if (client_ecdsa_capable != ctx.is_ecdsa_) {
      return false;
    }

    auto action = ocspStapleAction(ctx, client_ocsp_capable);
    if (action == OcspStapleAction::Fail) {
      return false;
    }

    selected_ctx = &ctx;
    ocsp_staple_action = action;
    return true;
  };

  auto select_from_map = [&]() -> void {
    if (server_name_it == server_names_map_.end()) {
      return;
    }
    const auto& pkey_types_map = server_name_it->second;
    for (const auto& entry : pkey_types_map) {
      if (selected(entry.second.get())) {
        break;
      }
    }
  };

  // Do SNI matching and pkey type matching if SNI exists.
  if (!sni.empty()) {
    // Match on exact server name, i.e. "www.example.com" for "www.example.com".
    server_name_it = server_names_map_.find(sni);
    select_from_map();

    if (selected_ctx == nullptr) {
      // Match on wildcard domain, i.e. ".example.com" for "www.example.com".
      // https://datatracker.ietf.org/doc/html/rfc6125#section-6.4
      size_t pos = sni.find('.', 1);
      if (pos < sni.size() - 1 && pos != std::string::npos) {
        absl::string_view wildcard = sni.substr(pos);
        server_name_it = server_names_map_.find(static_cast<std::string>(wildcard));
      }
      select_from_map();
    }
  }
  // Do pkey type matching if SNI does Not exist or no ctx matched for SNI.
  if (selected_ctx == nullptr) {
    // Fallback on first certificate.
    selected_ctx = &tls_contexts_[0];
    ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
    for (const auto& ctx : tls_contexts_) {
      if (selected(ctx)) {
        break;
      }
    }
  }
```

Reply:
Yes, I think something like this is the correct way to do it. This prevents a partial match/selection on exact name from blocking a full match/selection on wildcard.

But to make that code snippet clearer, I think you should pass the iterator as a parameter to the lambda, and have the lambda only capture `this`.

Reply:
OK, will update the PR.